src/async_fuse/memfs/cache/memory_cache/write_back_task.rs

```diff
@@ -32,205 +33,294 @@ pub enum Command {
-) where
-    P: EvictPolicy<BlockCoordinate> + Send + Sync,
-    S: Storage + Send + Sync,
-{
-    let file_level_pending_blocks = pending_blocks.remove(&ino);
-    lru_queue.retain_with_order(|&BlockCoordinate(i, _)| i != ino);
-
-    if let Some(file_level_pending_blocks) = file_level_pending_blocks {
-        for (block_id, (block, sender)) in file_level_pending_blocks {
-            cache.backend().store(ino, block_id, block).await;
-            sender.send(()).unwrap_or_else(|()| {
-                warn!("The receiver of pending block is closed unexpectedly.");
-            });
+    /// Calculate a new period for write back task's interval.
+    fn calculate_period(&self) -> Duration {
+        let pending_length = self.lru_queue.len();
+
+        if pending_length > self.command_queue_limit.overflow_div(4).overflow_mul(3) {
+            // If the number of pending blocks is greater than 75% of `command_queue_limit`,
+            // the new period should be changed to `1/4`.
+            self.interval / 4
+        } else if pending_length > self.command_queue_limit.overflow_div(2) {
+            // If the number of pending blocks is greater than 50% of `command_queue_limit`,
+            // the new period should be changed to `1/2`.
+            self.interval / 2
+        } else {
+            // Otherwise, restore the basic period.
+            self.interval
         }
     }
-}
 
-/// Flush all files, write blocks of them to backend immediately.
-async fn flush_all<P, S>(
-    lru_queue: &mut LinkedHashSet<BlockCoordinate>,
-    pending_blocks: &mut HashMap<INum, HashMap<BlockId, (Block, oneshot::Sender<()>)>>,
-    cache: Arc<MemoryCache<P, S>>,
-    sender: oneshot::Sender<()>,
-) where
-    P: EvictPolicy<BlockCoordinate> + Send + Sync,
-    S: Storage + Send + Sync,
-{
-    lru_queue.clear();
+    /// Flush a pending block to the backend
+    ///
+    /// # Panics
+    /// If this function is called out of the context of a tokio runtime,
+    /// it will panic.
+    fn flush_a_block(&mut self, cache: &Arc<MemoryCache<P, S>>)
+    where
+        P: EvictPolicy<BlockCoordinate> + Send + Sync + 'static,
+        S: Storage + Send + Sync + 'static,
+    {
+        if let Some(BlockCoordinate(ino, block_id)) = self.lru_queue.pop_front() {
+            if let Some((block, sender)) = self
+                .pending_blocks
+                .get_mut(&ino)
+                .and_then(|file_level_pending_blocks| file_level_pending_blocks.remove(&block_id))
+            {
+                let handle = tokio::spawn(write_back_one_block(
+                    Arc::clone(cache),
+                    ino,
+                    block_id,
+                    block,
+                    sender,
+                ));
+                self.task_queue.push_back(handle);
+            }
+        }
+    }
+
+    /// Flush a file, write blocks of it to backend immediately.
+    async fn flush_file(&mut self, cache: Arc<MemoryCache<P, S>>, ino: INum)
```

Comment:
This function is quite similar with the later one, can we refactor them?

Reply:
Ther are similar now, but when we introduce `TaskManager` here, they will perform different behaviours: `flush_file` spawns tasks to flush blocks, and submits them into the `TaskManager`; while `flush_all` spawns those tasks and await them immediately, then send a signal to the caller in `MemoryCache` via `sender`. That's because `flush_all` is usually called once when shutdown, it's not able to spawn task with `TaskManager`.